File: tests/viewTransform.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import {
  actualSizeViewTransform,
  createViewTransform,
  effectiveScale,
  fitScale,
  fitViewTransform,
  imageToViewport,
  panByViewportDelta,
  resizeViewTransform,
  stepViewZoom,
  viewportToImage,
  zoomAtViewportPoint,
} from "../src/domain/viewTransform";

const close = (actual: number, expected: number, tolerance = 1e-9) =>
  assert.ok(Math.abs(actual - expected) <= tolerance, `${actual} != ${expected}`);

test("calculates contain fit scale without aspect ratio distortion", () => {
  close(fitScale({ width: 406, height: 720 }, { width: 1000, height: 600 }), 600 / 720);
  close(fitScale({ width: 1920, height: 1080 }, { width: 600, height: 1000 }), 600 / 1920);
});

test("round trips image and viewport coordinates", () => {
  const transform = createViewTransform({ width: 406, height: 720 }, { width: 1000, height: 600 });
  const imagePoint = { x: 123.25, y: 456.75 };
  const roundTrip = viewportToImage(transform, imageToViewport(transform, imagePoint));
  close(roundTrip.x, imagePoint.x);
  close(roundTrip.y, imagePoint.y);
});

test("keeps the cursor image point anchored while zooming", () => {
  const transform = createViewTransform({ width: 1000, height: 1000 }, { width: 800, height: 600 });
  const anchor = { x: 600, y: 200 };
  const before = viewportToImage(transform, anchor);
  const zoomed = zoomAtViewportPoint(transform, 3, anchor);
  const after = viewportToImage(zoomed, anchor);
  close(after.x, before.x);
  close(after.y, before.y);
});

test("clamps zoom and pan while centering undersized axes", () => {
  const transform = createViewTransform({ width: 406, height: 720 }, { width: 1000, height: 600 });
  const zoomed = zoomAtViewportPoint(transform, 100, { x: 500, y: 300 });
  assert.equal(zoomed.zoom, 10);
  const panned = panByViewportDelta(zoomed, { x: 100000, y: -100000 });
  const scale = effectiveScale(panned);
  assert.ok(panned.center.x >= panned.viewportSize.width / (2 * scale));
  assert.ok(panned.center.y <= panned.imageSize.height - panned.viewportSize.height / (2 * scale));
  const fit = fitViewTransform(panned);
  assert.deepEqual(fit.center, { x: 203, y: 360 });
  assert.equal(fit.zoom, 1);
});

test("preserves the viewed image center across resize and frame reuse", () => {
  const transform = panByViewportDelta(
    zoomAtViewportPoint(createViewTransform({ width: 1000, height: 800 }, { width: 800, height: 600 }), 2, { x: 400, y: 300 }),
    { x: -80, y: 40 },
  );
  const resized = resizeViewTransform(transform, { width: 1200, height: 700 });
  close(resized.center.x, transform.center.x);
  close(resized.center.y, transform.center.y);
  assert.equal(resized.zoom, transform.zoom);
  assert.equal(resized, resizeViewTransform(resized, resized.viewportSize));
});

test("repeated inverse zooms do not drift at the viewport center", () => {
  let transform = createViewTransform({ width: 1000, height: 800 }, { width: 800, height: 600 });
  const anchor = { x: 400, y: 300 };
  for (let index = 0; index < 100; index += 1) {
    transform = zoomAtViewportPoint(transform, 2, anchor);
    transform = zoomAtViewportPoint(transform, 1, anchor);
  }
  close(transform.center.x, 500);
  close(transform.center.y, 400);
  assert.equal(transform.zoom, 1);
});

test("steps zoom by fixed ten percentage points instead of multiplying", () => {
  const anchor = { x: 400, y: 300 };
  let transform = createViewTransform({ width: 1000, height: 800 }, { width: 800, height: 600 });
  transform = stepViewZoom(transform, 1, anchor);
  assert.equal(transform.zoom, 1.1);
  transform = stepViewZoom(transform, 1, anchor);
  assert.equal(transform.zoom, 1.2);
  transform = stepViewZoom(transform, -1, anchor);
  assert.equal(transform.zoom, 1.1);
  transform = stepViewZoom(fitViewTransform(transform), -1, anchor);
  assert.equal(transform.zoom, 1);
  const maximum = zoomAtViewportPoint(transform, 10, anchor);
  assert.equal(stepViewZoom(maximum, 1, anchor), maximum);
});

test("keeps fixed-step cursor anchor stable without round-trip drift", () => {
  const anchor = { x: 400, y: 300 };
  let transform = createViewTransform({ width: 1000, height: 800 }, { width: 800, height: 600 });
  const before = viewportToImage(transform, anchor);
  for (let index = 0; index < 100; index += 1) {
    transform = stepViewZoom(transform, 1, anchor);
    transform = stepViewZoom(transform, -1, anchor);
  }
  const after = viewportToImage(transform, anchor);
  close(after.x, before.x);
  close(after.y, before.y);
  assert.equal(transform.zoom, 1);
});

test("sets original pixel size while preserving the viewed center", () => {
  const transform = panByViewportDelta(
    zoomAtViewportPoint(createViewTransform({ width: 1000, height: 800 }, { width: 800, height: 600 }), 2, { x: 400, y: 300 }),
    { x: -50, y: 25 },
  );
  const actual = actualSizeViewTransform(transform);
  close(effectiveScale(actual), 1);
  close(actual.center.x, transform.center.x);
  close(actual.center.y, transform.center.y);
});

File: android/app/src/test/java/com/snowberried/ctcinereviewer/render/ViewTransformTest.kt
package com.snowberried.ctcinereviewer.render

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ViewTransformTest {
    @Test
    fun `pinch keeps the image point under its anchor`() {
        val initial = createViewTransform(ViewSize(1_000f, 500f), ViewSize(400f, 400f))
        val anchor = ViewPoint(300f, 200f)
        val anchoredImagePoint = viewportToImage(initial, anchor)

        val zoomed = zoomViewTransform(initial, 2f, anchor)

        assertPointEquals(anchor, imageToViewport(zoomed, anchoredImagePoint))
        assertEquals(2f, zoomed.zoom, EPSILON)
        assertFalse(zoomed.isFit)
    }

    @Test
    fun `pan clamps every large axis and centers a smaller axis`() {
        val fit = createViewTransform(ViewSize(1_000f, 500f), ViewSize(400f, 400f))
        val zoomed = zoomViewTransform(fit, 3f, ViewPoint(200f, 200f))

        val topLeftLimit = panViewTransform(zoomed, ViewPoint(10_000f, 10_000f))
        val bottomRightLimit = panViewTransform(zoomed, ViewPoint(-20_000f, -20_000f))

        val firstPlacement = viewPlacement(topLeftLimit)
        val secondPlacement = viewPlacement(bottomRightLimit)
        assertEquals(0f, firstPlacement.left, EPSILON)
        assertEquals(0f, firstPlacement.top, EPSILON)
        assertEquals(400f, secondPlacement.left + secondPlacement.width, EPSILON)
        assertEquals(400f, secondPlacement.top + secondPlacement.height, EPSILON)

        val smallerAxis = createViewTransform(ViewSize(1_000f, 500f), ViewSize(400f, 400f))
        assertEquals(250f, smallerAxis.center.y, EPSILON)
    }

    @Test
    fun `fit reset restores contain fit zoom and center`() {
        val initial = createViewTransform(ViewSize(1_000f, 500f), ViewSize(400f, 400f))
        val changed = panViewTransform(
            zoomViewTransform(initial, 4f, ViewPoint(250f, 220f)),
            ViewPoint(-80f, 45f),
        )

        val reset = fitViewTransform(changed)

        assertTrue(reset.isFit)
        assertEquals(1f, reset.zoom, EPSILON)
        assertPointEquals(ViewPoint(500f, 250f), reset.center)
        assertEquals(ViewPlacement(0f, 100f, 400f, 200f), viewPlacement(reset))
    }

    @Test
    fun `viewport resize preserves manual scale then clamps safely`() {
        val initial = createViewTransform(ViewSize(1_000f, 500f), ViewSize(400f, 400f))
        val changed = panViewTransform(
            zoomViewTransform(initial, 2f, ViewPoint(200f, 200f)),
            ViewPoint(-80f, 0f),
        )
        val scaleBefore = effectiveScale(changed)

        val resized = resizeViewTransform(changed, ViewSize(300f, 500f))

        assertEquals(scaleBefore, effectiveScale(resized), EPSILON)
        assertEquals(changed.center.x, resized.center.x, EPSILON)
        assertEquals(250f, resized.center.y, EPSILON)
        val placement = viewPlacement(resized)
        assertTrue(placement.left <= EPSILON)
        assertTrue(placement.left + placement.width >= 300f - EPSILON)
        assertEquals(50f, placement.top, EPSILON)
        assertEquals(450f, placement.top + placement.height, EPSILON)
    }

    private fun assertPointEquals(expected: ViewPoint, actual: ViewPoint) {
        assertEquals(expected.x, actual.x, EPSILON)
        assertEquals(expected.y, actual.y, EPSILON)
    }

    private companion object {
        const val EPSILON = 0.001f
    }
}
